File: apprefurbished/src/main/java/com/androidcollider/easyfin/common/ui/fragments/common/CommonFragmentAddEdit.java
package com.androidcollider.easyfin.common.ui.fragments.common;

import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.TypedValue;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.androidcollider.easyfin.R;
import com.androidcollider.easyfin.accounts.add_edit.AddAccountFragment;
import com.androidcollider.easyfin.common.app.App;
import com.androidcollider.easyfin.common.managers.ui.dialog.DialogManager;
import com.androidcollider.easyfin.common.ui.MainActivity;
import com.androidcollider.easyfin.common.ui.fragments.FrgNumericDialog;

import javax.inject.Inject;

/**
 * @author Ihor Bilous
 */

public abstract class CommonFragmentAddEdit extends CommonFragment {

    @Inject
    DialogManager dialogManager;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((App) getActivity().getApplication()).getComponent().inject(this);
    }

    @Override
    public String getTitle() {
        return getString(R.string.app_name);
    }

    protected void setToolbar() {
        ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (actionBar != null) {
            ViewGroup actionBarLayout = (ViewGroup) getActivity().getLayoutInflater().inflate(
                    R.layout.save_close_buttons_toolbar, null);

            ActionBar.LayoutParams layoutParams = new ActionBar.LayoutParams(
                    ActionBar.LayoutParams.MATCH_PARENT,
                    ActionBar.LayoutParams.MATCH_PARENT);

            actionBar.setDisplayShowTitleEnabled(false);
            actionBar.setDisplayHomeAsUpEnabled(false);
            actionBar.setDisplayShowCustomEnabled(true);
            actionBar.setCustomView(actionBarLayout, layoutParams);

            Toolbar parent = (Toolbar) actionBarLayout.getParent();
            parent.setContentInsetsAbsolute(0, 0);

            Button btnSave = (Button) actionBarLayout.findViewById(R.id.btnToolbarSave);
            Button btnClose = (Button) actionBarLayout.findViewById(R.id.btnToolbarClose);

            btnSave.setOnClickListener(v -> handleSaveAction());

            btnClose.setOnClickListener(v -> finish());
        }
    }

    protected void setTVTextSize(TextView textView, String s, int min, int max) {
        int length = s.length();
        if (length > min && length <= max)
            textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 30);
        else if (length > max)
            textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 24);
        else
            textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 36);
    }

    protected void openNumericDialog(String initialValue) {
        Bundle args = new Bundle();
        args.putString("value", initialValue);

        DialogFragment numericDialog = new FrgNumericDialog();
        numericDialog.setTargetFragment(this, 1);
        numericDialog.setArguments(args);
        numericDialog.show(getActivity().getSupportFragmentManager(), "numericDialog");
    }

    protected void showDialogNoAccount(String message, boolean withFinish) {
        MainActivity activity = (MainActivity) getActivity();
        if (activity != null) {
            dialogManager.showNoAccountsDialog(
                    activity,
                    message,
                    (dialog, which) -> goToAddAccount(withFinish),
                    (dialog, which) -> finish()
            );
        }
    }

    private void goToAddAccount(boolean withFinish) {
        if (withFinish) finish();

        AddAccountFragment addAccountFragment = new AddAccountFragment();
        Bundle arguments = new Bundle();
        arguments.putInt("mode", 0);
        addAccountFragment.setArguments(arguments);

        addFragment(addAccountFragment);
    }

    protected abstract void handleSaveAction();
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/common/ui/fragments/common/CommonFragmentAddEdit.kt
package com.androidcollider.easyfin.common.ui.fragments.common

import android.os.Bundle
import android.util.TypedValue
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.core.os.bundleOf
import androidx.fragment.app.setFragmentResultListener
import androidx.navigation.fragment.findNavController
import com.androidcollider.easyfin.R
import com.androidcollider.easyfin.accounts.list.AccountsFragment.Companion.MODE
import com.androidcollider.easyfin.common.app.App
import com.androidcollider.easyfin.common.managers.ui.dialog.DialogManager
import com.androidcollider.easyfin.common.ui.fragments.NumericDialogFragment.Companion.INPUT_VALUE
import com.androidcollider.easyfin.common.ui.fragments.NumericDialogFragment.Companion.NUMERIC_DIALOG_REQUEST_KEY
import com.androidcollider.easyfin.common.ui.fragments.NumericDialogFragment.Companion.OUTPUT_VALUE
import com.androidcollider.easyfin.common.utils.setSafeOnClickListener
import javax.inject.Inject

/**
 * @author Ihor Bilous
 */
abstract class CommonFragmentAddEdit : CommonFragment() {

    private lateinit var btnSave: Button
    private lateinit var btnClose: Button

    @Inject
    lateinit var dialogManager: DialogManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity?.application as App).component?.inject(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        btnSave = view.findViewById(R.id.btnSave)
        btnClose = view.findViewById(R.id.btnClose)
        btnSave.setSafeOnClickListener { handleSaveAction() }
        btnClose.setSafeOnClickListener { findNavController().navigateUp() }

        setFragmentResultListener(NUMERIC_DIALOG_REQUEST_KEY) { _, bundle ->
            updateAmount(bundle.getString(OUTPUT_VALUE, "0,00"))
        }
    }

    override val title: String
        get() = getString(R.string.app_name)

    protected fun setTVTextSize(textView: TextView, s: String, min: Int, max: Int) {
        val length = s.length
        when {
            length in (min + 1)..max ->
                textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 30f)
            length > max -> textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 24f)
            else -> textView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 36f)
        }
    }

    protected fun openNumericDialog(initialValue: String) {
        findNavController().navigate(
            R.id.numericDialogFragment,
            bundleOf(
                INPUT_VALUE to initialValue
            )
        )
    }

    protected fun showDialogNoAccount(message: String, withFinish: Boolean) {
        activity?.let {
            dialogManager.showNoAccountsDialog(
                it,
                message,
                { goToAddAccount(withFinish) }) { findNavController().navigateUp() }
        }
    }

    private fun goToAddAccount(withFinish: Boolean) {
        val navController = findNavController()
        if (withFinish) navController.navigateUp()
        navController.navigate(
            R.id.addAccountFragment,
            bundleOf(MODE to 0)
        )
    }

    protected abstract fun handleSaveAction()
    protected abstract fun updateAmount(amount: String)
}